[UnifiedPDF] Implement pinch-zooming (macOS)

hortont424 opened the pull request:
#### 1d980c4592b05f24494e5fd86640b4289f2e01e5
<pre>
[UnifiedPDF] Implement pinch-zooming (macOS)
<a href="https://bugs.webkit.org/show_bug.cgi?id=265760">https://bugs.webkit.org/show_bug.cgi?id=265760</a>
&lt;<a href="https://rdar.apple.com/problem/114832109">rdar://problem/114832109</a>&gt;

Reviewed by Simon Fraser.

Implement pinch zooming (not HUD zooming or etc.) for the UnifiedPDFPlugin for macOS.

* Source/WebKit/UIProcess/ViewGestureController.cpp:
(WebKit::ViewGestureController::didCollectGeometryForMagnificationGesture):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::didBeginMagnificationGesture):
(WebKit::WebPageProxy::didEndMagnificationGesture):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
Let the WebContent process know when we start a gesture (we already send the matching &quot;end&quot;).

* Source/WebKit/WebProcess/Plugins/PDF/PDFPlugin.h:
* Source/WebKit/WebProcess/Plugins/PDF/PDFPlugin.mm:
(WebKit::PDFPlugin::installPDFDocument):
(WebKit::PDFPlugin::deviceScaleFactorChanged):
(WebKit::PDFPlugin::geometryDidChange):
(WebKit::PDFPlugin::setPageScaleFactor):
(WebKit::PDFPlugin::notifyContentScaleFactorChanged):
(WebKit::PDFPlugin::contentsScaleFactorChanged): Deleted.
(WebKit::PDFPlugin::handlesPageScaleFactor const): Deleted.
Two small changes to the legacy PDFPlugin:
- rename contentsScaleFactorChanged to deviceScaleFactorChanged since that&apos;s what it&apos;s used for.
- Adopt the new `origin` parameter on setPageScaleFactor instead of using the &quot;last mouse position&quot;.

This makes us match Preview and Web Content behavior when using the keyboard to zoom
(around the center of the view, not around the mouse!), and avoids the need to store
the last mouse position in UnifiedPDFPlugin (there remains one use in legacy PDFPlugin).

* Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.h:
(WebKit::PDFPluginBase::deviceScaleFactorChanged):
(WebKit::PDFPluginBase::didBeginMagnificationGesture):
(WebKit::PDFPluginBase::didEndMagnificationGesture):
(WebKit::PDFPluginBase::contentsScaleFactorChanged): Deleted.
* Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.mm:
(WebKit::PDFPluginBase::handlesPageScaleFactor const):
* Source/WebKit/WebProcess/Plugins/PluginView.cpp:
(WebKit::PluginView::topContentInsetDidChange):
(WebKit::PluginView::didBeginMagnificationGesture):
(WebKit::PluginView::didEndMagnificationGesture):
(WebKit::PluginView::setPageScaleFactor):
(WebKit::PluginView::setDeviceScaleFactor):
(WebKit::PluginView::pageScaleFactorDidChange): Deleted.
* Source/WebKit/WebProcess/Plugins/PluginView.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::setTextZoomFactor):
(WebKit::WebPage::setPageZoomFactor):
(WebKit::WebPage::setPageAndTextZoomFactors):
(WebKit::WebPage::scalePage):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/mac/WebPageMac.mm:
(WebKit::WebPage::didBeginMagnificationGesture):
(WebKit::WebPage::didEndMagnificationGesture):
Plumb beginning and end of magnification gestures, and the gesture origin, to the plugin.

* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.h:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
(WebKit::UnifiedPDFPlugin::updateLayerHierarchy):
Mark our root layer as &quot;the one that applies the page scale&quot; if we&apos;re the whole main frame.

(WebKit::UnifiedPDFPlugin::paintContents):
(WebKit::UnifiedPDFPlugin::pageScaleFactor const):
(WebKit::UnifiedPDFPlugin::didBeginMagnificationGesture):
(WebKit::UnifiedPDFPlugin::didEndMagnificationGesture):
Keep track of when we&apos;re inside a magnification gesture, and tickle the layers
when the gesture ends.

(WebKit::UnifiedPDFPlugin::setPageScaleFactor):
Scale the contents layer with page scale, and counter-scroll to keep the
gesture origin in the same position.

For now, until we have async painting, we avoid updating the tiles until
the gesture ends.

(WebKit::UnifiedPDFPlugin::documentSize const):
(WebKit::UnifiedPDFPlugin::contentsSize const):
Split these two; documentSize is unscaled (scale=1); contentsSize is the
ScrollableArea contents size (which must have the scale applied).

Canonical link: <a href="https://commits.webkit.org/271501@main">https://commits.webkit.org/271501@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b0429d70df7edd1bdb1c3a6019e9437e6f7296df

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/28594 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/7239 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/29977 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/31241 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/26117 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/9352 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/4609 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/31241 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/28864 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/5985 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/24596 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/31241 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/5357 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/32560 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/26185 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/26042 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/32560 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/5320 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/3489 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/32560 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/6926 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/6844 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/5781 "Built successfully") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/5839 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.h:
```diff
@@ -142,6 +148,9 @@ class UnifiedPDFPlugin final : public PDFPluginBase, public WebCore::GraphicsLay
     WebCore::ScrollingNodeID m_scrollingNodeID { 0 };
+
+    bool m_inMagnificationGesture { false };
+    float m_scaleFactor { 1 };
```
Please flip for better padding.

smfr commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -282,6 +284,67 @@
+    m_rootLayer->noteDeviceOrPageScaleFactorChangedIncludingDescendants();
+}
+
+void UnifiedPDFPlugin::setPageScaleFactor(double scale, std::optional<WebCore::IntPoint> origin)
```
page scale is a `float` in most places.

hortont424 replied:
This method is not new.

smfr commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -282,6 +284,67 @@
+    options.originalScrollDelta = delta;
+    page->protectedScrollingCoordinator()->requestScrollToPosition(*this, newPosition, options);
+
+    scheduleRenderingUpdate();
```
Do you really need to call this?

hortont424 replied:
Good question. I touched the layer so I assumed so? But I will check, this has been here since early versions of the patch.

hortont424 committed:
[UnifiedPDF] Implement pinch-zooming (macOS)
https://bugs.webkit.org/show_bug.cgi?id=265760
<rdar://problem/114832109>

Reviewed by Simon Fraser.

Implement pinch zooming (not HUD zooming or etc.) for the UnifiedPDFPlugin for macOS.

* Source/WebKit/UIProcess/ViewGestureController.cpp:
(WebKit::ViewGestureController::didCollectGeometryForMagnificationGesture):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::didBeginMagnificationGesture):
(WebKit::WebPageProxy::didEndMagnificationGesture):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
Let the WebContent process know when we start a gesture (we already send the matching "end").

* Source/WebKit/WebProcess/Plugins/PDF/PDFPlugin.h:
* Source/WebKit/WebProcess/Plugins/PDF/PDFPlugin.mm:
(WebKit::PDFPlugin::installPDFDocument):
(WebKit::PDFPlugin::deviceScaleFactorChanged):
(WebKit::PDFPlugin::geometryDidChange):
(WebKit::PDFPlugin::setPageScaleFactor):
(WebKit::PDFPlugin::notifyContentScaleFactorChanged):
(WebKit::PDFPlugin::contentsScaleFactorChanged): Deleted.
(WebKit::PDFPlugin::handlesPageScaleFactor const): Deleted.
Two small changes to the legacy PDFPlugin:
- rename contentsScaleFactorChanged to deviceScaleFactorChanged since that's what it's used for.
- Adopt the new `origin` parameter on setPageScaleFactor instead of using the "last mouse position".

This makes us match Preview and Web Content behavior when using the keyboard to zoom
(around the center of the view, not around the mouse!), and avoids the need to store
the last mouse position in UnifiedPDFPlugin (there remains one use in legacy PDFPlugin).

* Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.h:
(WebKit::PDFPluginBase::deviceScaleFactorChanged):
(WebKit::PDFPluginBase::didBeginMagnificationGesture):
(WebKit::PDFPluginBase::didEndMagnificationGesture):
(WebKit::PDFPluginBase::contentsScaleFactorChanged): Deleted.
* Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.mm:
(WebKit::PDFPluginBase::handlesPageScaleFactor const):
* Source/WebKit/WebProcess/Plugins/PluginView.cpp:
(WebKit::PluginView::topContentInsetDidChange):
(WebKit::PluginView::didBeginMagnificationGesture):
(WebKit::PluginView::didEndMagnificationGesture):
(WebKit::PluginView::setPageScaleFactor):
(WebKit::PluginView::setDeviceScaleFactor):
(WebKit::PluginView::pageScaleFactorDidChange): Deleted.
* Source/WebKit/WebProcess/Plugins/PluginView.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::setTextZoomFactor):
(WebKit::WebPage::setPageZoomFactor):
(WebKit::WebPage::setPageAndTextZoomFactors):
(WebKit::WebPage::scalePage):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/mac/WebPageMac.mm:
(WebKit::WebPage::didBeginMagnificationGesture):
(WebKit::WebPage::didEndMagnificationGesture):
Plumb beginning and end of magnification gestures, and the gesture origin, to the plugin.

* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.h:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
(WebKit::UnifiedPDFPlugin::updateLayerHierarchy):
Mark our root layer as "the one that applies the page scale" if we're the whole main frame.

(WebKit::UnifiedPDFPlugin::paintContents):
(WebKit::UnifiedPDFPlugin::pageScaleFactor const):
(WebKit::UnifiedPDFPlugin::didBeginMagnificationGesture):
(WebKit::UnifiedPDFPlugin::didEndMagnificationGesture):
Keep track of when we're inside a magnification gesture, and tickle the layers
when the gesture ends.

(WebKit::UnifiedPDFPlugin::setPageScaleFactor):
Scale the contents layer with page scale, and counter-scroll to keep the
gesture origin in the same position.

For now, until we have async painting, we avoid updating the tiles until
the gesture ends.

(WebKit::UnifiedPDFPlugin::documentSize const):
(WebKit::UnifiedPDFPlugin::contentsSize const):
Split these two; documentSize is unscaled (scale=1); contentsSize is the
ScrollableArea contents size (which must have the scale applied).

Canonical link: https://commits.webkit.org/271501@main
